test/test_transforms_ad_rules.jl

```diff
@@ -0,0 +1,101 @@
+using SpeedyWeather
+using EnzymeTestUtils, Enzyme
+import EnzymeTestUtils: test_approx
+using FiniteDifferences
+import FiniteDifferences: j′vp, grad, central_fdm
+import AbstractFFTs
+
+grid_types = [FullGaussianGrid, OctahedralClenshawGrid] # one full and one reduced grid 
+
+# currenlty there's an issue with EnzymeTestUtils not being able to work with structs with undefined fields like FFT plans
+# https://github.com/EnzymeAD/Enzyme.jl/issues/1992
+# This is a very hacky workaround 
+function EnzymeTestUtils.test_approx(x::AbstractFFTs.Plan, y::AbstractFFTs.Plan, msg; kwargs...)
+    EnzymeTestUtils.@test_msg "$msg: types must match" typeof(x) == typeof(y)
+    names = fieldnames(typeof(x))[1:end-1] # exclude pinv field (which is the last field)
+    if isempty(names)
+        EnzymeTestUtils.@test_msg msg x == y
+    else
+        for k in names
+            if k isa Symbol && hasproperty(x, k)
+                msg_new = "$msg: ::$(typeof(x)).$k"
+            else
+                msg_new = "$msg: getfield(::$(typeof(x)), $k)"
+            end
+            EnzymeTestUtils.test_approx(getfield(x, k), getfield(y, k), msg_new; kwargs...)
+        end
+    end
+    return nothing
+end 
+
+@testset "SpeedyTransforms: AD Rules" begin
+
+    @testset "_fourier! Enzyme rules" begin
+        
+        @testset "reverse rule" begin
+            for grid_type in grid_types
+
+                spectral_grid = SpectralGrid(Grid=grid_type)
+                S = SpectralTransform(spectral_grid)
+                grid = rand(spectral_grid.Grid{spectral_grid.NF}, spectral_grid.nlat_half, spectral_grid.nlayers)
+                f_north = S.scratch_memory_north
+                f_south = S.scratch_memory_south
+
+                # not working currenlty, the test is stuck 
+                #test_reverse(SpeedyWeather.SpeedyTransforms._fourier!, Const, (f_north, Duplicated), (f_south, Duplicated), (grid, Duplicated), (S, Const))
+            end
+        end
+    end 
+
+    @testset "Complete Transform Enzyme" begin
+        # make a high level finite difference test of the whole transform
+        # can't use Enzyme or ChainRule Test tools for tests for that
+        for grid_type in grid_types
+
+            spectral_grid = SpectralGrid(Grid=grid_type)
+
+            # forwards 
+            S = SpectralTransform(spectral_grid)
+            dS = deepcopy(S)
+            grid = rand(spectral_grid.Grid{spectral_grid.NF}, spectral_grid.nlat_half, spectral_grid.nlayers)
+            dgrid = zero(grid)
+            specs = zeros(LowerTriangularArray{Complex{spectral_grid.NF}}, spectral_grid.trunc+2, spectral_grid.trunc+1, spectral_grid.nlayers)
+            
+            # seed
+            dspecs = zero(specs)
+            fill!(dspecs, 1+1im)
```

Comment:
What does "seed" mean in this context and why is it 1+i?

Reply:
We do a reverse mode differentiation here. So you propagate the value of the output in reverse direction to determine the sensitivity of output wrt its inputs. The seed is an example output so to say. If you set it to one, you do actually compute directly the gradient of the function. 

As I wrote above in the big post, differentiating complex numbers is a bit of special topics in itself. In our case we treat the imaginary and real part kind of like separate numbers, so setting both to one made sense to me intuitively. But differentiating complex numbers is something we should still think about a bit.

Reply:
Yes please! Just checking, because I think of 
```julia
julia> z = exp(im*π/4)
0.7071067811865476 + 0.7071067811865475im

julia> abs(z)
1.0
```
more as complex "1" than `1 + 1im` because it's actually a unit vector